ami/models/components/unet/commons.py

```diff
@@ -0,0 +1,146 @@
+        """A residual block for Unet Components.
+
+        Args:
+            in_channels (int):
+                Input num of channels.
+            emb_channels (int):
+                Input embedding num of channels.
+            out_channels (int):
+                Output num of channels.
+        """
+        super().__init__()
+        self.in_layers = nn.Sequential(
+            nn.GroupNorm(num_groups=32, num_channels=in_channels),
+            nn.SiLU(),
+            nn.Conv2d(in_channels, out_channels, 3, padding=1),
+        )
+        self.emb_layers = nn.Sequential(
+            nn.SiLU(),
+            nn.Linear(
+                emb_channels,
+                2 * out_channels,
+            ),
+        )
+        self.out_layers = nn.Sequential(
+            nn.GroupNorm(num_groups=32, num_channels=out_channels),
+            nn.SiLU(),
+            nn.Conv2d(out_channels, out_channels, 3, padding=1),
+        )
+        # initialize conv layer as zeros
+        for p in self.out_layers[-1].parameters():
+            p.detach().zero_()
+
+        self.skip_connection: nn.Module = (
+            nn.Identity() if out_channels == in_channels else nn.Conv2d(in_channels, out_channels, 1)
+        )
+
+    def forward(self, x: torch.Tensor, emb: torch.Tensor) -> torch.Tensor:
+        """Apply the block to input Tensor, conditioned on emb.
+
+        Args:
+            x (torch.Tensor):
+                Input features.
+                (shape: [batch_size, in_channels, height, width])
+            emb (torch.Tensor):
+                Input Embeddings for conditioning.
+                (shape: [batch_size, emb_channels])
+        Returns:
+            torch.Tensor:
+                Output features.
+                (shape: [batch_size, out_channels, height, width])
+        """
+        h = self.in_layers(x)
+        emb_out = self.emb_layers(emb).type(h.dtype)
+        while len(emb_out.shape) < len(h.shape):
+            emb_out = emb_out[..., None]
+        # FiLM
+        out_norm, out_rest = self.out_layers[0], self.out_layers[1:]
+        scale, shift = torch.chunk(emb_out, 2, dim=1)
+        h = out_norm(h) * (1 + scale) + shift
+        h = out_rest(h)
+        return self.skip_connection(x) + h
+
+
+# To Do:
+# Temporarily use this implementation of the original paper,
+# but it will be changed to use nn.MultiheadAttention in Unet components.
+class AttentionBlock(nn.Module):
+    """An attention block that allows spatial positions to attend to each
+    other.
+
+    Originally ported from here, but adapted to the N-d case.
+    https://github.com/hojonathanho/diffusion/blob/1e0dceb3b3495bbe19116a5e1b3596cd0706c543/diffusion_tf/models/unet.py#L66.
+    """
+
+    def __init__(
+        self,
+        channels: int,
+        num_heads: int,
+    ) -> None:
+        super().__init__()
+        self.channels = channels
+        self.num_heads = num_heads
+        self.norm = nn.GroupNorm(num_groups=32, num_channels=channels)
```

Comment:
ここもコンストラクタで指定できるようにしよう。
```suggestion
        self.norm = nn.GroupNorm(num_groups=num_groups, num_channels=channels)
```